### PR TITLE
Fix Issue 17474 - non-property being treated as a property [in Assign…

### PR DESCRIPTION
**Pull request by UplinkCoder**
…ment]

Since this is very counter-intuitive behavior I'd like to deprecate it.

### CODE REVIEW

**Comment by UplinkCoder** [comment]
@WalterBright @andralex any objections to this ?

**Comment by andralex** [comment]
This seems to also disallow this:

```d
ref int* getTheIntPtr();
getTheIntPtr = null; // deprecated by this PR
```

Is that correct?

**Comment by UplinkCoder** [comment]
@andralex no.
the deprecation only trigger when the property behavior would have been invoked.
since there is no argument in your `getThisIntPtr`
it won't trigger the deprecation.

File: changelog/deprecateNonPropertyAssignment.dd [hunk 1]
```diff
@@ -0,0 +1,16 @@
+the property assignment to non-@property functions has been deprecated.
```
Comment (andralex) [inline changelog/deprecateNonPropertyAssignment.dd]: I think we should do this: `Properties with exactly one defaulted value cannot be assigned to.` Is that correct?

File: changelog/deprecateNonPropertyAssignment.dd [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+Since it can result it can result in fairly unintutive errors
```
Comment (andralex) [inline changelog/deprecateNonPropertyAssignment.dd]: Replace with: `Rationale: in an assignment, it is ambiguous whether the property should be invoked with no argument and the result be assigned to, or the property should be invoked with one argument.`

File: src/ddmd/root/array.d [hunk 3]
```diff
@@ -199,12 +199,12 @@ public:
 struct BitArray
 {
 nothrow:
-    size_t length() const pure
+    @property size_t length() const pure
```
Comment (andralex) [inline src/ddmd/root/array.d]: Hmmmm... this code should not be broken because it is not ambiguous. If you have `size_t length();` and `void length(size_t)` it is unambiguous in all contexts what the get and set mean.

Ambiguity occurs when a property has a default value and returns a ref.

File: test/runnable/property.d [hunk 4]
```diff
@@ -1,4 +1,10 @@
-
+/*
+TEST_OUTPUT:
+---
+runnable/property.d(24): Deprecation: assignment to non-property function `int bar(int value)` is deprecated.
+runnable/property.d(47): Deprecation: assignment to non-property function `void prop(int v)` is deprecated.
+---
+*/
```
Comment (andralex) [inline test/runnable/property.d]: The tests in this file should still pass.

**Review by andralex — CHANGES_REQUESTED** [review]
This is good work but deprecates a bit more than needed. I think we should (at least for now) deprecate properties with defaulted values. cc @WalterBright

**Comment by UplinkCoder** [comment]
Ambiguity occurs, when the function has a default value, and does return a type where assignment is possible with the same type the default argument had.
Think opAssign or alias this.

**Comment by UplinkCoder** [comment]
I really want to avoid such special casing though ...
Can we agree to deprecate the setter-property behavior for functions which are not @property and do not return void ?

File: changelog/deprecateNonPropertyAssignment.dd (hunk 1 above)
Comment (UplinkCoder) [inline changelog/deprecateNonPropertyAssignment.dd]: such special casing will lead us a c++-like forest of rules and well special cases

**Comment by UplinkCoder** [comment]
@andralex I made the criterion for the deprecation stricter.
I cannot make it much narrower as then it will probably take considerable time to check.

**Comment by andralex** [comment]
We are looking for simple and clear rules indeed. How about this: "Properties must have either exactly zero or one parameter. If present, the parameter cannot be defaulted."

**Comment by UplinkCoder** [comment]
@andralex thinking about it some more I think its acceptable to deprecate property assignment (setter behavior) for non-property functions with default argument.
 
PR has been amended.

**Comment by UplinkCoder** [comment]
jenkins is failing because of a regression introduced by now calling dtors transitively. and has nothing todo with this pr.

File: src/ddmd/root/array.d (hunk 3 above)
Comment (UplinkCoder) [inline src/ddmd/root/array.d]: I'd like to leave the `@property` here because it is one.
Any other blockers @andralex

**Comment by andralex** [comment]
I can't follow from the conversation what works and what doesn't. 

One great thing to do for this PR is a complete passing test with things that work, and a number of tests required to fail, each with one thing that doesn't work. That would give us a good picture of what the behavior is.

We also need changes to the spec wording. I can do that after I see the tests.

cc @WalterBright

File: src/ddmd/root/array.d (hunk 3 above)
Comment (andralex) [inline src/ddmd/root/array.d]: So I understand this is a stylistic change. The problem with it being mixed in this PR is that now I worry the old code would no longer work.

Stylistic changes are best left for a separate PR where introducing regressions is not a problem. FWIW syle-wise we want to have as few property annotations as possible, not as many. So you'll have resistance from me if you tried to add property annotations without necessity.

**Comment by UplinkCoder** [comment]
@andralex I added the failing case as test/fail_compilation/issue17474.d
which fails because of the function assigned to is not annotated `@property` and has a default argument.
If it were annotated property we would allow the ambiguous behavior.

tldr; the deprecation does only affect the one ambiguous case in issue 17474.
Everything else just works as expected which can be seen by the test-suite passing without modification.

**Comment by JinShil** [comment]
@UplinkCoder Are you still interested in pursuing this?  If not, I'd like to adopt it.

**Comment by AndrewEdwards** [comment]
@UplinkCoder - Need to know if you are still pursuing this. If you are, please update.

**Comment by UplinkCoder** [comment]
I don't think this will make it
